File: internal/api/handler/setting.go
package handler

import (
	"github.com/labstack/echo/v4"

	"clepsydra/internal/api"
	"clepsydra/internal/service"
	"clepsydra/internal/workday"
)

// Setting 设置相关接口
type Setting struct {
	settingSvc  *service.Setting
	holidaySvc  *service.HolidaySvc
}

// NewSetting 构建设置 handler
func NewSetting(settingSvc *service.Setting, holidaySvc *service.HolidaySvc) *Setting {
	return &Setting{
		settingSvc:  settingSvc,
		holidaySvc:  holidaySvc,
	}
}

// All GET /api/settings
func (h *Setting) All(c echo.Context) error {
	values, err := h.settingSvc.All(c.Request().Context())
	if err != nil {
		return api.Fail(c, err)
	}

	return api.OK(c, values)
}

// Update PUT /api/settings
func (h *Setting) Update(c echo.Context) error {
	var req struct {
		Values map[string]string `json:"values"`
	}
	if err := c.Bind(&req); err != nil {
		return api.Fail(c, service.ErrBadRequest("参数错误"))
	}

	err := h.settingSvc.Update(c.Request().Context(), req.Values)
	if err != nil {
		return api.Fail(c, err)
	}

	return api.OK(c, nil)
}

// Holidays GET /api/holidays?year=2026
func (h *Setting) Holidays(c echo.Context) error {
	year := c.QueryParam("year")

	holidays, err := h.holidaySvc.List(c.Request().Context(), year)
	if err != nil {
		return api.Fail(c, err)
	}

	return api.OK(c, holidays)
}

// SaveHolidays PUT /api/holidays
func (h *Setting) SaveHolidays(c echo.Context) error {
	var req struct {
		Entries []workday.Entry `json:"entries"`
	}
	if err := c.Bind(&req); err != nil {
		return api.Fail(c, service.ErrBadRequest("参数错误"))
	}

	err := h.holidaySvc.Save(c.Request().Context(), req.Entries)
	if err != nil {
		return api.Fail(c, err)
	}

	return api.OK(c, nil)
}

// DeleteHoliday DELETE /api/holidays/:date
func (h *Setting) DeleteHoliday(c echo.Context) error {
	date := c.Param("date")

	err := h.holidaySvc.Delete(c.Request().Context(), date)
	if err != nil {
		return api.Fail(c, err)
	}

	return api.OK(c, nil)
}

File: dashboard/apps/web-antdv-next/src/api/setting.ts
import { requestClient } from '#/api/request';

/** 查询全部设置项（仅超级管理员） */
export function fetchSettings() {
  return requestClient.get<Api.Setting.Values>('/api/settings');
}

/** 批量更新设置项（仅超级管理员） */
export function updateSettings(values: Api.Setting.Values): Promise<void> {
  return requestClient.put('/api/settings', { values });
}

/** 查询节假日列表，year 为空表示不筛选（仅超级管理员） */
export function fetchHolidays(year?: string) {
  return requestClient.get<Api.Setting.Holiday[]>('/api/holidays', {
    params: year ? { year } : undefined,
  });
}

/** 批量保存节假日，按日期覆盖更新（仅超级管理员） */
export function saveHolidays(
  entries: Api.Setting.HolidayEntry[],
): Promise<void> {
  return requestClient.put('/api/holidays', { entries });
}

/** 删除指定日期的节假日记录（仅超级管理员） */
export function deleteHoliday(date: string): Promise<void> {
  return requestClient.delete(`/api/holidays/${date}`);
}
